type_info.go

```diff
@@ -0,0 +1,244 @@
+package duckdb
+
+/*
+#include <duckdb.h>
+*/
+import "C"
+
+import (
+	"reflect"
+	"runtime"
+	"unsafe"
+)
+
+type baseTypeInfo struct {
+	// The data type.
+	t Type
+	// This slice serves different purposes.
+	// - STRUCT child names.
+	// - ENUM dictionary names.
+	names []string
+	// The  width of DECIMAL types.
+	width uint8
+	// The scale of DECIMAL types.
+	scale uint8
+}
+
+// TypeInfo contains all information to work with DuckDB types.
+type TypeInfo struct {
+	baseTypeInfo
+	// The child type information of nested types.
+	childTypes []TypeInfo
+}
+
+type vectorTypeInfo struct {
+	baseTypeInfo
+	// The dictionary for ENUM types.
+	dict map[string]uint32
+}
+
+// PrimitiveTypeInfo returns primitive type information.
+func PrimitiveTypeInfo(t Type) (TypeInfo, error) {
+	name, inMap := unsupportedTypeToStringMap[t]
+	if inMap {
+		return TypeInfo{baseTypeInfo: baseTypeInfo{t: t}}, getError(errAPI, unsupportedTypeError(name))
+	}
+
+	switch t {
+	case TYPE_DECIMAL:
+		return TypeInfo{}, getError(errAPI, tryOtherFuncError(funcName(DecimalTypeInfo)))
+	case TYPE_ENUM:
+		return TypeInfo{}, getError(errAPI, tryOtherFuncError(funcName(EnumTypeInfo)))
+	case TYPE_LIST:
+		return TypeInfo{}, getError(errAPI, tryOtherFuncError(funcName(ListTypeInfo)))
+	case TYPE_STRUCT:
+		return TypeInfo{}, getError(errAPI, tryOtherFuncError(funcName(StructTypeInfo)))
+	case TYPE_MAP:
+		return TypeInfo{}, getError(errAPI, tryOtherFuncError(funcName(MapTypeInfo)))
+	}
+
+	return TypeInfo{baseTypeInfo: baseTypeInfo{t: t}}, nil
+}
+
+// DecimalTypeInfo returns DECIMAL type information.
+func DecimalTypeInfo(width uint8, scale uint8) TypeInfo {
+	return TypeInfo{
+		baseTypeInfo: baseTypeInfo{
+			t:     TYPE_DECIMAL,
+			width: width,
+			scale: scale,
+		},
+	}
+}
+
+// EnumTypeInfo returns ENUM type information.
+func EnumTypeInfo(dict []string) (TypeInfo, error) {
+	if len(dict) == 0 {
+		return TypeInfo{}, getError(errAPI, errEmptySlice)
+	}
+
+	typeInfo := TypeInfo{baseTypeInfo: baseTypeInfo{t: TYPE_ENUM}}
+	typeInfo.names = make([]string, len(dict))
+	copy(typeInfo.names, dict)
+	return typeInfo, nil
+}
+
+// ListTypeInfo returns LIST type information.
+func ListTypeInfo(childInfo TypeInfo) (TypeInfo, error) {
+	t := TypeInfo{
+		baseTypeInfo: baseTypeInfo{t: TYPE_LIST},
+		childTypes:   make([]TypeInfo, 1),
+	}
+
+	if childInfo.t == TYPE_INVALID {
+		return t, getError(errAPI, errInvalidChildType)
+	}
+	t.childTypes[0] = childInfo
+	return t, nil
+}
+
+// StructTypeInfo returns STRUCT type information.
+func StructTypeInfo(typeInfos []TypeInfo, names []string) (TypeInfo, error) {
```

Comment:
Whats the benefit of this over `[]StructField`? I think its a better API enforcing equal type infos and names. But its very minor, just a nitpick

Reply:
That's a good point; I'll add it. 👍